framework/include/gpu/base/GPUMap.h

```diff
@@ -0,0 +1,84 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "GPUArray.h"
+
+#include <memory>
+#include <map>
+
+template <typename T1, typename T2>
```

Comment:
Docstrings (at least for the class). Move function definitions outside of class declaration. Needs to be namespaced